Add the new map editor.

pchote opened the pull request:
A revamp of #4496.

Mailaender committed:
Add menu plumbing for the new map editor.

pchote committed:
Add Remove support to TypeDictionary.

pchote committed:
Add methods to IOccupySpaceInfo.

pchote committed:
Add new map editor UI.

pchote committed:
Add the world components of the new editor.

Mailaender commented:
Regressions compared to https://github.com/OpenRA/OpenRA/pull/4496 include:

* mousewheel zoom is gone
* mousewheel actor rotation is not working
* infantry won't animate
* cloaked units do not show the cloaked palette (to check the radius of the cloak generator, maybe solves this with range circle)
* low power is not indicated anymore with disabled overlays
* infantry is not doing the idle animation while everything else does
* tooltips don't work neither at the battlefield nor in the selection palettes

Improvements I spotted so far:
* Crates appear in the actor palette

Mailaender commented:
Crash when placing `mpspawn` in the ts mod:

```
Exception of type `System.Collections.Generic.KeyNotFoundException`: The given key was not present in the dictionary.
  at System.Collections.Generic.Dictionary`2[OpenRA.Player,OpenRA.Traits.Stance].get_Item (OpenRA.Player key) [0x000a2] in /home/abuild/rpmbuild/BUILD/mono-3.12.1/mcs/class/corlib/System.Collections.Generic/Dictionary.cs:148 
  at OpenRA.Player.IsAlliedWith (OpenRA.Player p) [0x0000e] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Player.cs:138 
  at OpenRA.Mods.Common.Traits.ConquestVictoryConditions+<Tick>c__AnonStorey0.<>m__0 (OpenRA.Player p) [0x00017] in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.Common/Traits/Player/ConquestVictoryConditions.cs:47 
  at System.Linq.Enumerable+<CreateWhereIterator>c__Iterator1E`1[OpenRA.Player].MoveNext () [0x00059] in /home/abuild/rpmbuild/BUILD/mono-3.12.1/mcs/class/System.Core/System.Linq/Enumerable.cs:3120 
  at System.Linq.Enumerable.Any[Player] (IEnumerable`1 source) [0x00024] in /home/abuild/rpmbuild/BUILD/mono-3.12.1/mcs/class/System.Core/System.Linq/Enumerable.cs:138 
  at OpenRA.Mods.Common.Traits.ConquestVictoryConditions.Tick (OpenRA.Actor self) [0x000d4] in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.Common/Traits/Player/ConquestVictoryConditions.cs:49 
  at OpenRA.World.<Tick>m__4 (TraitPair`1 x) [0x0000e] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/World.cs:307 
  at (wrapper delegate-invoke) System.Action`1<OpenRA.TraitPair`1<OpenRA.Traits.ITick>>:invoke_void_T (OpenRA.TraitPair`1<OpenRA.Traits.ITick>)
  at OpenRA.WorldUtils.DoTimed[TraitPair`1] (IEnumerable`1 e, System.Action`1 a, System.String text) [0x00022] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/WorldUtils.cs:92 
  at OpenRA.World.Tick () [0x0013d] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/World.cs:307 
  at OpenRA.Game.InnerLogicTick (OpenRA.Network.OrderManager orderManager) [0x001bf] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:466 
  at OpenRA.Game.LogicTick () [0x0004e] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:490 
  at OpenRA.Game.Loop () [0x000d4] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:606 
  at OpenRA.Game.Run () [0x00042] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:646 
  at OpenRA.Program.Run (System.String[] args) [0x00011] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Support/Program.cs:111 
  at OpenRA.Program.Main (System.String[] args) [0x00050] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Support/Program.cs:39
```

pchote commented:
> Regressions compared to #4496 include:

These are all intentional changes except for the tooltips, which will be a simple fix.

Mailaender commented:
> * tooltips don't work neither at the battlefield nor in the selection palettes
> * INTERIOR boxes appear on DESERT maps (with a glitchy palette error)
> * click drag 1x1 tiles no longer works which e.g. makes painting oceans very annoying.

and the crash should be fixed again prior to merging. I assume that won't be hard.

pchote commented:
> INTERIOR boxes appear on DESERT maps (with a glitchy palette error)

This isn't a regression - the RA mod rules are just sloppy with their editor filtering (this affects more than just boxes on desert).  I have a commit that explores using multiple inheritance to clean up, correct, and drastically reduce the duplicated map decoration rules, but that is outside the scope of this particular PR.

penev92 commented:
Would be nice if D2k's resource layer is drawn under the terrain grid.
Also it would be very nice if you could bind the zoom to a hotkey. Or even better - to the mouse scroll.

pchote commented:
Zoom is bound to the zoom hotkey (`.` by default).  I am against binding zoom to the scrollwheel because scroll != zoom and our zoom is in discrete steps, while the scrollwheel / trackpad is continuous.

x-a-n-a-x commented:
What about using PageUp and PageDown keys for zooming in and out ?

pchote committed:
Make sure that ScreenMap is initialized before other traits.

obrakmann commented:
Just a couple of things while doing some quick testing:

* Exiting and discarding all changes still leaves an orphaned `temp.oramap` behind.
* Loading the desert shellmap crashes with `Exception of type 'System.InvalidOperationException': TypeDictionary does not contain instance of type 'OpenRA.Traits.IQuantizeBodyOrientationInfo'`

Mailaender commented:
> Would be nice if D2k's resource layer is drawn under the terrain grid.

I already fixed it by simply re-ordering the traits. Looks like a recent regression.

pchote commented:
Resource rendering is completely different now.

pchote commented:
Now depends on #8124 and #8125.

pchote committed:
Don’t crash if IQBO isn’t present on an actor.

pchote committed:
Add ClickThrough property to ContainerWidget.

pchote committed:
Render terrain grid above actors, too.

pchote commented:
Fixed all the relevant comments.

Mailaender commented:
Middle mouse scrolling does not work in the editor. This regressed a few times in my #4496. You should be able to steal the fix. It is mostly about widget re-ordering, inheritance and splitting.

Mailaender commented:
The tiles still flicker in the TS build palette (see Bridges for example).

pchote committed:
Force the first template variant in TTPW.

pchote commented:
Fixed both.

pchote commented:
Rebased.

Mailaender commented:
I created some maps on the cnc mod placing some player spawns, tiles and actors for testing, but the game will never show them. Not even on game restart.

I tried to open them in the legacy OpenRA.Editor.exe, but that also crashes. Apparently with every map so that could be a regression of https://github.com/OpenRA/OpenRA/pull/8125.

```
System.InvalidOperationException: TypeDictionary does not contain instance of type `OpenRA.Traits.PlayerColorPaletteInfo`
  at OpenRA.Primitives.TypeDictionary.Get[PlayerColorPaletteInfo] () [0x00089] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Primitives/TypeDictionary.cs:63 
  at OpenRA.Editor.Surface.GetPaletteForPlayerInner (System.String name) [0x00036] in /home/matthias/Entwicklung/OpenRA/OpenRA.Editor/Surface.cs:394 
  at OpenRA.Exts.GetOrAdd[String,ColorPalette] (System.Collections.Generic.Dictionary`2 d, System.String k, System.Func`2 createFn) [0x00013] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Exts.cs:129 
  at OpenRA.Primitives.Cache`2[System.String,System.Drawing.Imaging.ColorPalette].get_Item (System.String key) [0x0000e] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Primitives/Cache.cs:35 
  at OpenRA.Editor.Surface.GetPaletteForPlayer (System.String player) [0x0002a] in /home/matthias/Entwicklung/OpenRA/OpenRA.Editor/Surface.cs:406 
  at OpenRA.Editor.Surface.GetPaletteForActor (OpenRA.ActorReference ar) [0x00021] in /home/matthias/Entwicklung/OpenRA/OpenRA.Editor/Surface.cs:415 
  at OpenRA.Editor.Surface.OnPaint (System.Windows.Forms.PaintEventArgs e) [0x0042f] in /home/matthias/Entwicklung/OpenRA/OpenRA.Editor/Surface.cs:467 
  at System.Windows.Forms.Control.WmPaint (System.Windows.Forms.Message& m) [0x0008a] in /home/abuild/rpmbuild/BUILD/mono-3.12.1/mcs/class/Managed.Windows.Forms/System.Windows.Forms/Control.cs:5245 
  at System.Windows.Forms.Control.WndProc (System.Windows.Forms.Message& m) [0x0015f] in /home/abuild/rpmbuild/BUILD/mono-3.12.1/mcs/class/Managed.Windows.Forms/System.Windows.Forms/Control.cs:5007 
  at System.Windows.Forms.Control+ControlWindowTarget.OnMessage (System.Windows.Forms.Message& m) [0x00000] in /home/abuild/rpmbuild/BUILD/mono-3.12.1/mcs/class/Managed.Windows.Forms/System.Windows.Forms/Control.cs:228 
  at System.Windows.Forms.Control+ControlNativeWindow.WndProc (System.Windows.Forms.Message& m) [0x00000] in /home/abuild/rpmbuild/BUILD/mono-3.12.1/mcs/class/Managed.Windows.Forms/System.Windows.Forms/Control.cs:209 
  at System.Windows.Forms.NativeWindow.WndProc (IntPtr hWnd, Msg msg, IntPtr wParam, IntPtr lParam) [0x0008e] in /home/abuild/rpmbuild/BUILD/mono-3.12.1/mcs/class/Managed.Windows.Forms/System.Windows.Forms/NativeWindow.cs:240
```

pchote commented:
That legacy editor crash is indeed a regression from #8125.  Could you please create a separate issue for that?

Mailaender commented:
Sure. https://github.com/OpenRA/OpenRA/issues/8142 I am still a bit confused why the ingame map editor seems to create maps that are invisible for the game.

pchote commented:
That will be a bug that I will need to fix.

pchote commented:
Updated to fix the invisible maps and @x-a-n-a-x's cursor glitch.

Phrohdoh commented on OpenRA.Mods.Common/Traits/World/EditorActorLayer.cs:
```diff
@@ -0,0 +1,268 @@
+			{
+				UpdateNeighbours(preview.Footprint);
+
+				if (reference.Type == "mpspawn")
```
The multiple references to this hardcoded string means it ought to be exposed to yaml, IMO.

pchote replied:
"mpspawn" is hardcoded in the core engine code, so this would only add confusion.

Mailaender replied:
You could make it a constant. I thought about it, too, but it should not be done here and now.

pchote replied:
We were going to unhardcode this as part of the implementation of #3359, but then the planned art wasn't finished.

Phrohdoh commented:
I can place structures on slopes. Is this intentional?

Mailaender commented:
We didn't want to limit the map authors creativity.

Phrohdoh commented:
Noticed nothing being broken and no unexpected behaviour.
Works in all mods for me. :+1:

LipkeGu commented:
Try to get it crashing... by placing quickly some tiles right in the bottom of the map :(
```
Red Alert Mod at Version {DEV_VERSION}
Operating System: Windows (Microsoft Windows NT 6.2.9200.0)
Runtime Version: .NET CLR 4.0.30319.34209
Exception of type `System.InvalidOperationException`: OpenGL Error: See graphics.log for details.
   bei OpenRA.Renderer.Sdl2.ErrorHandler.CheckGlError() in d:\OpenRA-ingame-map-editor\OpenRA-ingame-map-editor\OpenRA.Renderer.Sdl2\ErrorHandler.cs:Zeile 47.
   bei OpenRA.Renderer.Sdl2.VertexBuffer`1.SetData(T[] data, Int32 start, Int32 length) in d:\OpenRA-ingame-map-editor\OpenRA-ingame-map-editor\OpenRA.Renderer.Sdl2\VertexBuffer.cs:Zeile 48.
   bei OpenRA.Graphics.TerrainRenderer.UpdateCell(CPos cell) in d:\OpenRA-ingame-map-editor\OpenRA-ingame-map-editor\OpenRA.Game\Graphics\TerrainRenderer.cs:Zeile 78.
   bei System.Action`1.Invoke(T obj)
   bei OpenRA.CellLayer`1.set_Item(CPos cell, T value) in d:\OpenRA-ingame-map-editor\OpenRA-ingame-map-editor\OpenRA.Game\Map\CellLayer.cs:Zeile 88.
   bei OpenRA.Mods.Common.Widgets.EditorTileBrush.HandleMouseInput(MouseInput mi)
   bei OpenRA.Mods.Common.Widgets.EditorViewportControllerWidget.HandleMouseInput(MouseInput mi)
   bei OpenRA.Widgets.Widget.HandleMouseInputOuter(MouseInput mi) in d:\OpenRA-ingame-map-editor\OpenRA-ingame-map-editor\OpenRA.Game\Widgets\Widget.cs:Zeile 354.
   bei OpenRA.Widgets.Widget.HandleMouseInputOuter(MouseInput mi) in d:\OpenRA-ingame-map-editor\OpenRA-ingame-map-editor\OpenRA.Game\Widgets\Widget.cs:Zeile 345.
   bei OpenRA.Widgets.Widget.HandleMouseInputOuter(MouseInput mi) in d:\OpenRA-ingame-map-editor\OpenRA-ingame-map-editor\OpenRA.Game\Widgets\Widget.cs:Zeile 345.
   bei OpenRA.Widgets.Widget.HandleMouseInputOuter(MouseInput mi) in d:\OpenRA-ingame-map-editor\OpenRA-ingame-map-editor\OpenRA.Game\Widgets\Widget.cs:Zeile 345.
   bei OpenRA.Widgets.Widget.HandleMouseInputOuter(MouseInput mi) in d:\OpenRA-ingame-map-editor\OpenRA-ingame-map-editor\OpenRA.Game\Widgets\Widget.cs:Zeile 345.
   bei OpenRA.Widgets.Ui.HandleInput(MouseInput mi) in d:\OpenRA-ingame-map-editor\OpenRA-ingame-map-editor\OpenRA.Game\Widgets\Widget.cs:Zeile 90.
   bei OpenRA.DefaultInputHandler.<>c__DisplayClass7.<OnMouseInput>b__6() in d:\OpenRA-ingame-map-editor\OpenRA-ingame-map-editor\OpenRA.Game\Input\InputHandler.cs:Zeile 49.
   bei OpenRA.Sync.CheckSyncUnchanged[T](World world, Func`1 fn) in d:\OpenRA-ingame-map-editor\OpenRA-ingame-map-editor\OpenRA.Game\Sync.cs:Zeile 182.
   bei OpenRA.DefaultInputHandler.OnMouseInput(MouseInput input) in d:\OpenRA-ingame-map-editor\OpenRA-ingame-map-editor\OpenRA.Game\Input\InputHandler.cs:Zeile 49.
   bei OpenRA.Renderer.Sdl2.Sdl2Input.PumpInput(IInputHandler inputHandler) in d:\OpenRA-ingame-map-editor\OpenRA-ingame-map-editor\OpenRA.Renderer.Sdl2\Sdl2Input.cs:Zeile 86.
   bei OpenRA.Renderer.Sdl2.Sdl2GraphicsDevice.PumpInput(IInputHandler inputHandler) in d:\OpenRA-ingame-map-editor\OpenRA-ingame-map-editor\OpenRA.Renderer.Sdl2\Sdl2GraphicsDevice.cs:Zeile 360.
   bei OpenRA.Graphics.Renderer.EndFrame(IInputHandler inputHandler) in d:\OpenRA-ingame-map-editor\OpenRA-ingame-map-editor\OpenRA.Game\Graphics\Renderer.cs:Zeile 156.
   bei OpenRA.Game.RenderTick() in d:\OpenRA-ingame-map-editor\OpenRA-ingame-map-editor\OpenRA.Game\Game.cs:Zeile 527.
   bei OpenRA.Game.Loop() in d:\OpenRA-ingame-map-editor\OpenRA-ingame-map-editor\OpenRA.Game\Game.cs:Zeile 628.
   bei OpenRA.Game.Run() in d:\OpenRA-ingame-map-editor\OpenRA-ingame-map-editor\OpenRA.Game\Game.cs:Zeile 646.
   bei OpenRA.Program.Run(String[] args) in d:\OpenRA-ingame-map-editor\OpenRA-ingame-map-editor\OpenRA.Game\Support\Program.cs:Zeile 111.
   bei OpenRA.Program.Main(String[] args) in d:\OpenRA-ingame-map-editor\OpenRA-ingame-map-editor\OpenRA.Game\Support\Program.cs:Zeile 39.
```
```GL Error: InvalidValue
   bei OpenRA.Renderer.Sdl2.ErrorHandler.CheckGlError()
   bei OpenRA.Renderer.Sdl2.VertexBuffer`1.SetData(T[] data, Int32 start, Int32 length)
   bei OpenRA.Graphics.TerrainRenderer.UpdateCell(CPos cell)
   bei System.Action`1.Invoke(T obj)
   bei OpenRA.CellLayer`1.set_Item(CPos cell, T value)
   bei OpenRA.Mods.Common.Widgets.EditorTileBrush.HandleMouseInput(MouseInput mi)
   bei OpenRA.Mods.Common.Widgets.EditorViewportControllerWidget.HandleMouseInput(MouseInput mi)
   bei OpenRA.Widgets.Widget.HandleMouseInputOuter(MouseInput mi)
   bei OpenRA.Widgets.Widget.HandleMouseInputOuter(MouseInput mi)
   bei OpenRA.Widgets.Widget.HandleMouseInputOuter(MouseInput mi)
   bei OpenRA.Widgets.Widget.HandleMouseInputOuter(MouseInput mi)
   bei OpenRA.Widgets.Widget.HandleMouseInputOuter(MouseInput mi)
   bei OpenRA.Widgets.Ui.HandleInput(MouseInput mi)
   bei OpenRA.DefaultInputHandler.<>c__DisplayClass7.<OnMouseInput>b__6()
   bei OpenRA.Sync.CheckSyncUnchanged[T](World world, Func`1 fn)
   bei OpenRA.DefaultInputHandler.OnMouseInput(MouseInput input)
   bei OpenRA.Renderer.Sdl2.Sdl2Input.PumpInput(IInputHandler inputHandler)
   bei OpenRA.Renderer.Sdl2.Sdl2GraphicsDevice.PumpInput(IInputHandler inputHandler)
   bei OpenRA.Graphics.Renderer.EndFrame(IInputHandler inputHandler)
   bei OpenRA.Game.RenderTick()
   bei OpenRA.Game.Loop()
   bei OpenRA.Game.Run()
   bei OpenRA.Program.Run(String[] args)
   bei OpenRA.Program.Main(String[] args)


OpenGL Information:
Vendor: NVIDIA Corporation
Renderer: GeForce GTX 550 Ti/PCIe/SSE2
GL Version: 4.5.0 NVIDIA 347.52
Shader Version: 4.50 NVIDIA
Available extensions:
GL_AMD_multi_draw_indirect GL_ARB_arrays_of_arrays GL_ARB_base_instance GL_ARB_blend_func_extended GL_ARB_buffer_storage GL_ARB_clear_buffer_object GL_ARB_clear_texture GL_ARB_clip_control GL_ARB_color_buffer_float GL_ARB_compatibility GL_ARB_compressed_texture_pixel_storage GL_ARB_conservative_depth GL_ARB_compute_shader GL_ARB_compute_variable_group_size GL_ARB_conditional_render_inverted GL_ARB_copy_buffer GL_ARB_copy_image GL_ARB_cull_distance GL_ARB_debug_output GL_ARB_depth_buffer_float GL_ARB_depth_clamp GL_ARB_depth_texture GL_ARB_derivative_control GL_ARB_direct_state_access GL_ARB_draw_buffers GL_ARB_draw_buffers_blend GL_ARB_draw_indirect GL_ARB_draw_elements_base_vertex GL_ARB_draw_instanced GL_ARB_enhanced_layouts GL_ARB_ES2_compatibility GL_ARB_ES3_compatibility GL_ARB_ES3_1_compatibility GL_ARB_explicit_attrib_location GL_ARB_explicit_uniform_location GL_ARB_fragment_coord_conventions GL_ARB_fragment_layer_viewport GL_ARB_fragment_program GL_ARB_fragment_program_shadow GL_ARB_fragment_shader GL_ARB_framebuffer_no_attachments GL_ARB_framebuffer_object GL_ARB_framebuffer_sRGB GL_ARB_geometry_shader4 GL_ARB_get_program_binary GL_ARB_get_texture_sub_image GL_ARB_gpu_shader5 GL_ARB_gpu_shader_fp64 GL_ARB_half_float_pixel GL_ARB_half_float_vertex GL_ARB_imaging GL_ARB_indirect_parameters GL_ARB_instanced_arrays GL_ARB_internalformat_query GL_ARB_internalformat_query2 GL_NV_internalformat_sample_query GL_ARB_invalidate_subdata GL_ARB_map_buffer_alignment GL_ARB_map_buffer_range GL_ARB_multi_bind GL_ARB_multi_draw_indirect GL_ARB_multisample GL_ARB_multitexture GL_ARB_occlusion_query GL_ARB_occlusion_query2 GL_ARB_pipeline_statistics_query GL_ARB_pixel_buffer_object GL_ARB_point_parameters GL_ARB_point_sprite GL_ARB_program_interface_query GL_ARB_provoking_vertex GL_ARB_robust_buffer_access_behavior GL_ARB_robustness GL_ARB_sample_shading GL_ARB_sampler_objects GL_ARB_seamless_cube_map GL_ARB_separate_shader_objects GL_ARB_shader_atomic_counters GL_ARB_shader_bit_encoding GL_ARB_shader_draw_parameters GL_ARB_shader_group_vote GL_ARB_shader_image_load_store GL_ARB_shader_image_size GL_ARB_shader_objects GL_ARB_shader_precision GL_ARB_query_buffer_object GL_ARB_shader_storage_buffer_object GL_ARB_shader_subroutine GL_ARB_shader_texture_image_samples GL_ARB_shader_texture_lod GL_ARB_shading_language_100 GL_ARB_shading_language_420pack GL_ARB_shading_language_include GL_ARB_shading_language_packing GL_ARB_shadow GL_ARB_sparse_buffer GL_ARB_sparse_texture GL_ARB_stencil_texturing GL_ARB_sync GL_ARB_tessellation_shader GL_ARB_texture_barrier GL_ARB_texture_border_clamp GL_ARB_texture_buffer_object GL_ARB_texture_buffer_object_rgb32 GL_ARB_texture_buffer_range GL_ARB_texture_compression GL_ARB_texture_compression_bptc GL_ARB_texture_compression_rgtc GL_ARB_texture_cube_map GL_ARB_texture_cube_map_array GL_ARB_texture_env_add GL_ARB_texture_env_combine GL_ARB_texture_env_crossbar GL_ARB_texture_env_dot3 GL_ARB_texture_float GL_ARB_texture_gather GL_ARB_texture_mirror_clamp_to_edge GL_ARB_texture_mirrored_repeat GL_ARB_texture_multisample GL_ARB_texture_non_power_of_two GL_ARB_texture_query_levels GL_ARB_texture_query_lod GL_ARB_texture_rectangle GL_ARB_texture_rg GL_ARB_texture_rgb10_a2ui GL_ARB_texture_stencil8 GL_ARB_texture_storage GL_ARB_texture_storage_multisample GL_ARB_texture_swizzle GL_ARB_texture_view GL_ARB_timer_query GL_ARB_transform_feedback2 GL_ARB_transform_feedback3 GL_ARB_transform_feedback_instanced GL_ARB_transform_feedback_overflow_query GL_ARB_transpose_matrix GL_ARB_uniform_buffer_object GL_ARB_vertex_array_bgra GL_ARB_vertex_array_object GL_ARB_vertex_attrib_64bit GL_ARB_vertex_attrib_binding GL_ARB_vertex_buffer_object GL_ARB_vertex_program GL_ARB_vertex_shader GL_ARB_vertex_type_10f_11f_11f_rev GL_ARB_vertex_type_2_10_10_10_rev GL_ARB_viewport_array GL_ARB_window_pos GL_ATI_draw_buffers GL_ATI_texture_float GL_ATI_texture_mirror_once GL_S3_s3tc GL_EXT_texture_env_add GL_EXT_abgr GL_EXT_bgra GL_EXT_bindable_uniform GL_EXT_blend_color GL_EXT_blend_equation_separate GL_EXT_blend_func_separate GL_EXT_blend_minmax GL_EXT_blend_subtract GL_EXT_compiled_vertex_array GL_EXT_Cg_shader GL_EXT_depth_bounds_test GL_EXT_direct_state_access GL_EXT_draw_buffers2 GL_EXT_draw_instanced GL_EXT_draw_range_elements GL_EXT_fog_coord GL_EXT_framebuffer_blit GL_EXT_framebuffer_multisample GL_EXTX_framebuffer_mixed_formats GL_EXT_framebuffer_multisample_blit_scaled GL_EXT_framebuffer_object GL_EXT_framebuffer_sRGB GL_EXT_geometry_shader4 GL_EXT_gpu_program_parameters GL_EXT_gpu_shader4 GL_EXT_multi_draw_arrays GL_EXT_packed_depth_stencil GL_EXT_packed_float GL_EXT_packed_pixels GL_EXT_pixel_buffer_object GL_EXT_point_parameters GL_EXT_polygon_offset_clamp GL_EXT_provoking_vertex GL_EXT_rescale_normal GL_EXT_secondary_color GL_EXT_separate_shader_objects GL_EXT_separate_specular_color GL_EXT_shader_image_load_store GL_EXT_shader_integer_mix GL_EXT_shadow_funcs GL_EXT_stencil_two_side GL_EXT_stencil_wrap GL_EXT_texture3D GL_EXT_texture_array GL_EXT_texture_buffer_object GL_EXT_texture_compression_dxt1 GL_EXT_texture_compression_latc GL_EXT_texture_compression_rgtc GL_EXT_texture_compression_s3tc GL_EXT_texture_cube_map GL_EXT_texture_edge_clamp GL_EXT_texture_env_combine GL_EXT_texture_env_dot3 GL_EXT_texture_filter_anisotropic GL_EXT_texture_integer GL_EXT_texture_lod GL_EXT_texture_lod_bias GL_EXT_texture_mirror_clamp GL_EXT_texture_object GL_EXT_texture_shared_exponent GL_EXT_texture_sRGB GL_EXT_texture_sRGB_decode GL_EXT_texture_storage GL_EXT_texture_swizzle GL_EXT_timer_query GL_EXT_transform_feedback2 GL_EXT_vertex_array GL_EXT_vertex_array_bgra GL_EXT_vertex_attrib_64bit GL_EXT_import_sync_object GL_IBM_rasterpos_clip GL_IBM_texture_mirrored_repeat GL_KHR_context_flush_control GL_KHR_debug GL_KHR_robust_buffer_access_behavior GL_KHR_robustness GL_KTX_buffer_region GL_NV_bindless_multi_draw_indirect GL_NV_bindless_multi_draw_indirect_count GL_NV_blend_equation_advanced GL_NV_blend_square GL_NV_compute_program5 GL_NV_conditional_render GL_NV_copy_depth_to_color GL_NV_copy_image GL_NV_depth_buffer_float GL_NV_depth_clamp GL_NV_draw_texture GL_NV_ES1_1_compatibility GL_NV_ES3_1_compatibility GL_NV_explicit_multisample GL_NV_fence GL_NV_float_buffer GL_NV_fog_distance GL_NV_fragment_program GL_NV_fragment_program_option GL_NV_fragment_program2 GL_NV_framebuffer_multisample_coverage GL_NV_geometry_shader4 GL_NV_gpu_program4 GL_NV_gpu_program4_1 GL_NV_gpu_program5 GL_NV_gpu_program5_mem_extended GL_NV_gpu_program_fp64 GL_NV_gpu_shader5 GL_NV_half_float GL_NV_light_max_exponent GL_NV_multisample_coverage GL_NV_multisample_filter_hint GL_NV_occlusion_query GL_NV_packed_depth_stencil GL_NV_parameter_buffer_object GL_NV_parameter_buffer_object2 GL_NV_path_rendering GL_NV_pixel_data_range GL_NV_point_sprite GL_NV_primitive_restart GL_NV_register_combiners GL_NV_register_combiners2 GL_NV_shader_atomic_counters GL_NV_shader_atomic_float GL_NV_shader_buffer_load GL_NV_shader_storage_buffer_object GL_NV_texgen_reflection GL_NV_texture_barrier GL_NV_texture_compression_vtc GL_NV_texture_env_combine4 GL_NV_texture_multisample GL_NV_texture_rectangle GL_NV_texture_shader GL_NV_texture_shader2 GL_NV_texture_shader3 GL_NV_transform_feedback GL_NV_transform_feedback2 GL_NV_uniform_buffer_unified_memory GL_NV_vertex_array_range GL_NV_vertex_array_range2 GL_NV_vertex_attrib_integer_64bit GL_NV_vertex_buffer_unified_memory GL_NV_vertex_program GL_NV_vertex_program1_1 GL_NV_vertex_program2 GL_NV_vertex_program2_option GL_NV_vertex_program3 GL_NVX_conditional_render GL_NVX_gpu_memory_info GL_NV_shader_thread_group GL_KHR_blend_equation_advanced GL_SGIS_generate_mipmap GL_SGIS_texture_lod GL_SGIX_depth_texture GL_SGIX_shadow GL_SUN_slice_accum GL_WIN_swap_hint WGL_EXT_swap_control 
```

cjshmyr commented:
Here's a crash case I found when editing a map in TS. To reproduce all you need to do was place this same cliff tile one tile above the previous like such until it crashes

![repro](https://cloud.githubusercontent.com/assets/566742/7602990/3427608a-f8e9-11e4-97e1-31a78f4aeb31.png)

```
Tiberian Sun Mod at Version {DEV_VERSION}
Operating System: Windows (Microsoft Windows NT 6.2.9200.0)
Runtime Version: .NET CLR 4.0.30319.0
Exception of type `System.IndexOutOfRangeException`: Index was outside the bounds of the array.
   at OpenRA.Mods.Common.Traits.TerrainGeometryOverlay.<>c__DisplayClass4.<RenderAfterWorld>b__0(WVec c)
   at System.Linq.Enumerable.WhereSelectArrayIterator`2.MoveNext()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at OpenRA.Mods.Common.Traits.TerrainGeometryOverlay.RenderAfterWorld(WorldRenderer wr, Actor self)
   at OpenRA.Graphics.WorldRenderer.Draw() in c:\Users\Curtis\OpenRA\OpenRA.Game\Graphics\WorldRenderer.cs:line 144
   at OpenRA.Game.RenderTick() in c:\Users\Curtis\OpenRA\OpenRA.Game\Game.cs:line 506
   at OpenRA.Game.Loop() in c:\Users\Curtis\OpenRA\OpenRA.Game\Game.cs:line 628
   at OpenRA.Game.Run() in c:\Users\Curtis\OpenRA\OpenRA.Game\Game.cs:line 646
   at OpenRA.Program.Run(String[] args) in c:\Users\Curtis\OpenRA\OpenRA.Game\Support\Program.cs:line 111
   at OpenRA.Program.Main(String[] args) in c:\Users\Curtis\OpenRA\OpenRA.Game\Support\Program.cs:line 39
```

Phrohdoh commented:
Obviously I didn't test enough odd-cases, thanks @cjshmyr and @LipkeGu. Removing my :+1:.

LipkeGu commented:
i would suggest, that we try to make the Map editor (as its own) nearest stable ... before merging...

pchote commented:
Closes #2939.

pchote commented:
Added additional checks to the tile brush to fix @cjshmyr's cliff-bootstrapping crash and @LipkeGu's outside-map crash.  Fixed a `// TODO: forbid bad values` in the original code to fix @LipkeGu's bogus map size crash.

LipkeGu commented:
restesting... :)

LipkeGu commented:
Crashes when more than one starting points are on the same location.

Reproduce case:
1. create a map
2. place 4 - 8 starting points on the same cell
3. Save the Map and load it in skirmish
4. Assign starting points to the AI (A = Player -> B: AI)
```
Red Alert Mod at Version {DEV_VERSION}
Operating System: Windows (Microsoft Windows NT 6.2.9200.0)
Runtime Version: .NET CLR 4.0.30319.34209
Exception of type `System.ArgumentException`: Ein Element mit dem gleichen Schlüssel wurde bereits hinzugefügt.
   bei System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   bei System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   bei System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, IEqualityComparer`1 comparer)
   bei System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector)
   bei OpenRA.Mods.Common.Widgets.Logic.LobbyUtils.GetSpawnOccupants(Session lobbyInfo, MapPreview preview)
   bei OpenRA.Mods.Common.Widgets.Logic.LobbyMapPreviewLogic.<>c__DisplayClass47.<.ctor>b__3()
   bei OpenRA.Mods.Common.Widgets.MapPreviewWidget.Draw()
   bei OpenRA.Widgets.Widget.DrawOuter() in d:\OpenRA-ingame-map-editor\OpenRA.Game\Widgets\Widget.cs:Zeile 411.
   bei OpenRA.Widgets.Widget.DrawOuter() in d:\OpenRA-ingame-map-editor\OpenRA.Game\Widgets\Widget.cs:Zeile 413.
   bei OpenRA.Widgets.Widget.DrawOuter() in d:\OpenRA-ingame-map-editor\OpenRA.Game\Widgets\Widget.cs:Zeile 413.
   bei OpenRA.Widgets.Widget.DrawOuter() in d:\OpenRA-ingame-map-editor\OpenRA.Game\Widgets\Widget.cs:Zeile 413.
   bei OpenRA.Widgets.Widget.DrawOuter() in d:\OpenRA-ingame-map-editor\OpenRA.Game\Widgets\Widget.cs:Zeile 413.
   bei OpenRA.Widgets.Widget.DrawOuter() in d:\OpenRA-ingame-map-editor\OpenRA.Game\Widgets\Widget.cs:Zeile 413.
   bei OpenRA.Widgets.Widget.DrawOuter() in d:\OpenRA-ingame-map-editor\OpenRA.Game\Widgets\Widget.cs:Zeile 413.
   bei OpenRA.Widgets.Ui.Draw() in d:\OpenRA-ingame-map-editor\OpenRA.Game\Widgets\Widget.cs:Zeile 77.
   bei OpenRA.Game.RenderTick() in d:\OpenRA-ingame-map-editor\OpenRA.Game\Game.cs:Zeile 517.
   bei OpenRA.Game.Loop() in d:\OpenRA-ingame-map-editor\OpenRA.Game\Game.cs:Zeile 628.
   bei OpenRA.Game.Run() in d:\OpenRA-ingame-map-editor\OpenRA.Game\Game.cs:Zeile 646.
   bei OpenRA.Program.Run(String[] args) in d:\OpenRA-ingame-map-editor\OpenRA.Game\Support\Program.cs:Zeile 111.
   bei OpenRA.Program.Main(String[] args) in d:\OpenRA-ingame-map-editor\OpenRA.Game\Support\Program.cs:Zeile 39.
```

pchote commented:
That isn't an editor problem.  See #8154.

LipkeGu commented:
is it able to check this already in the Editor?

cjshmyr commented:
1) I get a crash when creating a new TS map
```
Tiberian Sun Mod at Version {DEV_VERSION}
Operating System: Windows (Microsoft Windows NT 6.2.9200.0)
Runtime Version: .NET CLR 4.0.30319.0
Exception of type `System.IndexOutOfRangeException`: Index was outside the bounds of the array.
   at OpenRA.Mods.Common.Traits.TerrainGeometryOverlay.<>c__DisplayClass4.<RenderAfterWorld>b__0(WVec c)
   at System.Linq.Enumerable.WhereSelectArrayIterator`2.MoveNext()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at OpenRA.Mods.Common.Traits.TerrainGeometryOverlay.RenderAfterWorld(WorldRenderer wr, Actor self)
   at OpenRA.Graphics.WorldRenderer.Draw() in c:\Users\Curtis\OpenRA\OpenRA.Game\Graphics\WorldRenderer.cs:line 144
   at OpenRA.Game.RenderTick() in c:\Users\Curtis\OpenRA\OpenRA.Game\Game.cs:line 506
   at OpenRA.Game.Loop() in c:\Users\Curtis\OpenRA\OpenRA.Game\Game.cs:line 628
   at OpenRA.Game.Run() in c:\Users\Curtis\OpenRA\OpenRA.Game\Game.cs:line 646
   at OpenRA.Program.Run(String[] args) in c:\Users\Curtis\OpenRA\OpenRA.Game\Support\Program.cs:line 111
   at OpenRA.Program.Main(String[] args) in c:\Users\Curtis\OpenRA\OpenRA.Game\Support\Program.cs:line 39

```

2) On a map, place two actors. Left click on one actor then hover over any other actor and you get a stale tooltip.
![tooltip](https://cloud.githubusercontent.com/assets/566742/7626055/66117fc4-f9be-11e4-9d6e-489daf72da59.png)

cjshmyr commented:
TD's WINTER appears to just be TEMPERAT again -- intended?

cjshmyr commented:
If this is a non issue or intended, okay.

Specific to the `Actors` tab in the editor there is inconsistency of the types of actors that can be placed.

| Mod | Tileset | Actors from these tilesets can be placed | Comment |
|-----|----------|------------------------------------------|----------------------------------|
| TD | DESERT | DESERT | :+1: |
|  | WINTER | WINTER, DESERT | Remove DESERT actors? |
|  | SNOW | WINTER, DESERT | Remove DESERT actors? |
|  | TEMPERAT | TEMPERAT, DESERT | Remove DESERT actors? |
|  | JUNGLE | JUNGLE, DESERT | Remove DESERT actors? |
| RA | SNOW | SNOW, INTERIOR, DESERT | Remove DESERT & INTERIOR actors? |
|  | INTERIOR | INTERIOR, TEMPERAT, DESERT | Remove TEMPERAT & DESERT actors? |
|  | TEMPERAT | TEMPERAT, DESERT | Remove DESERT actors? |
|  | DESERT | DESRT, INTERIOR | Remove INTERIOR actors? |
| TS | SNOW | SNOW, TEMPERAT | Remove TEMPERAT actors? |
|  | TEMPERAT | TEMPERAT | :+1: |
| D2K | ARRAKIS | ARRAKIS | :) |

cjshmyr commented:
4) RA/TD behavior was consistent with the current editor, but in TS under the Tiles tab we can place tiles from either `TEMPERAT` or `SNOW` regardless of what we've chosen -- was this intended?

Mailaender commented:
@cjshmyr I already filed https://github.com/OpenRA/OpenRA/pull/8130 for that.

cjshmyr commented:
I played with #8130 and it seemed to fix the 4th item I posted, as for the 3rd item is the intent to let us place those actors in the editor but just fix the incorrect palette?

Mailaender commented:
It doesn't do anything with tiles, it just (re-)adds tileset filters as we abandoned the filter by file ending.

cjshmyr commented:
Okay. Still some oddities but I will post those in #8130

pchote commented:
* TS map creation crash: fixed by adding `Map.Contains` checks before dirtying resource tiles.
* TS tooltip thing: this happens because the actor selection area is set based on the size of the sprites (which is huge for TS).  We can't use the defined selection bounds because the code surrounding that is a giant mess (#8120, #6987, etc).
* Inconsistent actors: The editor's just doing what the mod is telling it to do, so outside the scope of this pr.
* Inconsistent tiles: The editor's just doing what the mod is telling it to do, so outside the scope of this pr.  Blame Westwood :)

LipkeGu commented:
@pchote : no it is already in the temperat.ini :/ (oil wreckage)

pchote commented:
Rebased.

obrakmann commented on mods/ra/chrome/editor.yaml:
```diff
@@ -0,0 +1,367 @@
+			Font: Bold
+			Key: escape
+		Button@SAVE:
+			X: 210
```
This button is misaligned on the X axis.

obrakmann commented:
Looks good to me overall. I'd still prefer it if the orphaned temp.oramap file could be removed when exiting the editor, but if the solution for that is too cumbersome can live it.

:+1: when button is aligned and the facings are corrected.

pchote committed:
Overhaul save panel.

pchote committed:
Fix create map panel geometry.

pchote commented:
Asking me to fix "simple" UI issues is always a dangerous thing.  Fixed the map creation dialog, and overhauled the save dialog.

penev92 commented:
OK I think we're ready to move to real-life tests with this.

LipkeGu commented:
An large Milestone is Done =) Yay xD

WolfGaming commented:
Congratz on a marvelous thing :)

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/510b8438a20f6c7ad00a2e648b3a0b70ce490871%5E...510b8438a20f6c7ad00a2e648b3a0b70ce490871)
